[RFC] Asynchronous notifications

jenswi-linaro opened the pull request:
This is a preview and RFC of the asynchronous notifications feature.

Note this is tested on QEMU virt (v7) together with the kernel patches in PR https://github.com/linaro-swg/linux/pull/93

I'm in particular interested in feedback on this from a SCMI point of view.
Other aspects to keep in mind is how this harmonizes with asynchronous notifications in FF-A 1.1

jforissier commented on core/include/kernel/notif.h:
```diff
@@ -0,0 +1,110 @@
+ * NOTIF_VALUE_MAX	    for mutex and condvar wait/wakeup
+ *
+ * Any value can be signalled with notif_send_sync() while only the ones
+ * <= CFG_CORE_ASYNC_NOTIF_MAX can be signalled with notif_send_async().
```
When calling `notif_send_sync()` with a value in the async range, is it still handled synchronously? (I suppose it is, I just want to make sure).

jenswi-linaro replied:
Yes, I'll update the description of notif_send_sync().

jforissier reviewed (commented):
A few comments already, although I have not looked at everything.

jenswi-linaro commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -452,6 +464,68 @@
+ * a2	Bit[0]: OPTEE_SMC_ASYNC_NOTIF_VALUE_VALID if the value in a1 is
+ *		valid, else 0 if no values where pending
+ * a2	Bit[1]: OPTEE_SMC_ASYNC_NOTIF_VALUE_PENDING if another value is
+ *		pending, else 0.
```
Called from a non-secure interrupt handler, see
https://github.com/jenswi-linaro/linux-1/blob/d84e6a4447f3d8b0e72fe3779ca8b28060b40abb/drivers/tee/optee/notif.c#L47-L57

jenswi-linaro commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -135,6 +135,12 @@
+ * 1. It must be able to handle eventual RPCs
+ * 2. Non-secure interrupts should not be masked
+ * 3. If asynchronous notifications has be negotiated successfully, then
+ *    asynchronous notifications should be unmasked during this call.
```
OK

jenswi-linaro commented on core/arch/arm/kernel/link_dummies_paged.c:
```diff
@@ -32,9 +32,3 @@ __thread_std_smc_entry(uint32_t a0 __unused, uint32_t a1 __unused,
-	 const void *sync_obj __unused, const char *fname __unused,
-	 int lineno  __unused)
-{
-}
```
OK

jenswi-linaro commented on core/include/kernel/notif.h:
```diff
@@ -0,0 +1,110 @@
+ */
+
+#if CFG_CORE_ASYNC_NOTIF_INTID
+#define NOTIF_ASYNC_VALUE_MAX		U(63)
```
Not really, they are slightly different things. In the Linux driver it's an assumption on the largest notification value that we assume we need to support **prior** to this patch. While here it's the largest asynchronous notification value we'll ever use.

FF-A 1.1 defines a kind of asynchronous notification value too, and this happens to be limited to maximum 63 so it should be a good match to this interface.

jenswi-linaro commented on mk/config.mk:
```diff
@@ -652,3 +652,9 @@ CFG_COMPAT_GP10_DES ?= y
 CFG_CORE_MAX_SYSCALL_RECURSION ?= 4
+
+# CFG_CORE_ASYNC_NOTIF_INTID is defined by the platform to some free
+# interrupt, larger than 31 to make it of the type SPI. Setting it to a
```
I'll update

jenswi-linaro commented:
Rebased on master on top the commits merged already. Split up a bit as requested.

jforissier commented on core/kernel/notif.c:
```diff
@@ -3,12 +3,217 @@
+
+	old_itr_status = cpu_spin_lock_xsave(&notif_lock);
+
+	/* Add ndrv at the end of the list */
+	nd = SLIST_FIRST(&notif_driver_head);
+	while (nd && SLIST_NEXT(nd, link))
+		nd = SLIST_NEXT(nd, link);
+	if (nd)
+		SLIST_INSERT_AFTER(nd, ndrv, link);
+	else
+		SLIST_INSERT_HEAD(&notif_driver_head, ndrv, link);
```
Why not insert at head? It shouldn't matter as far as I can tell.

jenswi-linaro replied:
You're right. I'll fix.

jforissier commented on core/include/kernel/notif.h:
```diff
@@ -11,19 +11,100 @@
+
+struct notif_driver {
+	void (*atomic_cb)(struct notif_driver *ndrv, enum notif_event ev);
+	void (*yielding_cb)(struct notif_driver *ndrv, enum notif_event ev);
```
Some comments would be welcome. For example on `atomic_cb`: `/* Called with exception and interrupts masked and notification spinlock held */`. Maybe also mention which values of `ev` are handled by which callback?

jenswi-linaro replied:
I'll fix.

jforissier reviewed (commented):
A couple more things. This looks quite useful to me, especially as a remember past discussions with people requesting that kind of functionality.

jenswi-linaro commented:
Update

etienne-lms commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -452,6 +465,71 @@
+ * Retrieve a value of notifications pended since the last call of this
+ * function.
+ *
+ * OP-TEE keeps a records of all posted values. When an interrupts is
```
s/records/record/
s/interrupts/interrupt/

etienne-lms commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -452,6 +465,71 @@
+ *
+ * OP-TEE keeps a records of all posted values. When an interrupts is
+ * received which indicates that there are posted values this function
+ * should be called until all pended values has been retrieved. When a
```
s/has/have/

etienne-lms commented on core/arch/arm/include/sm/optee_smc.h:
```diff
@@ -452,6 +465,71 @@
+ * should be called until all pended values has been retrieved. When a
+ * value is retrieved it's cleared from the record in secure world.
+ *
+ * It is expected that this function is called from and interrupt handler
```
s/and/an/

etienne-lms commented on core/arch/arm/tee/entry_fast.c:
```diff
@@ -240,6 +263,19 @@ void __tee_entry_fast(struct thread_smc_args *args)
 
+	case OPTEE_SMC_ENABLE_ASYNC_NOTIF:
+		if (notif_async_is_enabled())
+			notif_deliver_atomic_event(NOTIF_EVENT_STARTED);
```
should set `args->a0`.

etienne-lms commented on core/include/kernel/notif.h:
```diff
@@ -0,0 +1,175 @@
+ * NOTIF_VALUE_MAX	    for mutex and condvar wait/wakeup
+ *
+ * Any value can be signalled with notif_send_sync() while only the ones
+ * <= CFG_CORE_ASYNC_NOTIF_MAX can be signalled with notif_send_async().
```
s/CFG_CORE_ASYNC_NOTIF_MAX/NOTIF_ASYNC_VALUE_MAX/

etienne-lms commented on core/include/kernel/notif.h:
```diff
@@ -0,0 +1,175 @@
+TEE_Result notif_wait(uint32_t value);
+
+/*
+ * Send an asynchronous value, not that it must be <= NOTIF_ASYNC_VALUE_MAX
```
s/not/note/

etienne-lms commented on core/include/kernel/notif.h:
```diff
@@ -0,0 +1,175 @@
+ * has been delivered again.
+ *
+ * Note that while a @NOTIF_EVENT_STOPPED is being delivered at the same
+ * time may a @NOTIF_EVENT_STARTED be delivered again so a driver is
```
maybe replace `a driver` with `non-secure world`?
(or did I misunderstood?)

jenswi-linaro replied:
A driver is an OP-TEE driver in this case.
I'll remove "shared" from "shared internal state" since I guess that's what put you on the wrong track.

etienne-lms commented on core/include/kernel/notif.h:
```diff
@@ -0,0 +1,175 @@
+}
+#endif
+
+TEE_Result notif_alloc_async_value(uint32_t *val);
```
s/`val`/`value`/g ?

etienne-lms commented on core/kernel/notif.c:
```diff
@@ -0,0 +1,224 @@
+
+	cpu_spin_unlock_xrestore(&notif_lock, old_itr_status);
+
+	if (bit < 0)
```
maybe a `DMSG()` trace here to help identifying the out-of-mem reporter?

jenswi-linaro replied:
I'm not sure that helps much. This is likely to happen when the drivers are initializing so there's bound some more interesting print or perhaps even a panic.

etienne-lms commented on core/kernel/notif.c:
```diff
@@ -0,0 +1,224 @@
+bool notif_async_is_started(void)
+{
+	uint32_t old_itr_status = 0;
+	bool ret;
```
`= false;`

etienne-lms commented on core/tee/entry_std.c:
```diff
@@ -561,7 +562,22 @@ uint32_t __tee_entry_std(struct optee_msg_arg *arg, uint32_t num_params)
+		if (notif_async_is_enabled())
+			notif_deliver_event(NOTIF_EVENT_DO_BOTTOM_HALF);
+		else
+			goto err;
```
why not `rv = OPTEE_SMC_RETURN_EBADCMD;` straight?

jenswi-linaro replied:
I thought it nice to get the "Unknown cmd..." print in such a case.

etienne-lms reviewed (commented):
only minor comments

jenswi-linaro commented:
Update

jforissier commented:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

jenswi-linaro commented:
Squashed and tags applied

jforissier commented:
@etienne-lms do you want to provide a review or ack tag?

etienne-lms commented:
Sorry. Yes, `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with a late minor comment for commit "core: add asynchronous notifications".

jenswi-linaro commented:
> Sorry. Yes, `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with a late minor comment for commit "core: add asynchronous notifications".

I'm sorry, which comment?

etienne-lms reviewed (commented):
oops! this one.

jenswi-linaro commented:
Addressed comment, applied tag and rebased on master. We should wait with the merging this until we have made some more progress with upstreaming the kernel patches.

jforissier commented:
> We should wait with the merging this until we have made some more progress with upstreaming the kernel patches.

Sure.

jenswi-linaro commented:
Rebased on master.
The Linux kernel pull request has been merged into the arm-soc tree, see https://git.kernel.org/pub/scm/linux/kernel/git/soc/soc.git/commit/?id=c6e5bdae04a3

We've now reached a point where we can merge this.

jforissier commented:
> We've now reached a point where we can merge this.

Absolutely, thanks!

jenswi-linaro commented:
Yes, and yes.
